### PR TITLE
add URL Dependencies support for module federation(consume shared module)

### PR DESCRIPTION
**Pull request by liulangyu90316**
For module federation shared, without `requiredVersion` configured, it will infer version base on package.json, as we can see in `ConsumeSharedPlugin.js`. 
But when we get requiredVersion from package.json, it just get dep / dev / peer / optional dependencies value, without consideration of some URL Dependencies cases.
For example, `"isarray": "git+https://github.com/juliangruber/isarray.git#v2.0.3"`, it's real version should be `v2.0.3`, but we can not get that. 
This pr will add a good support for URL Dependencies with versions.

**What kind of change does this PR introduce?**
Feature, add URL Dependencies support for module federation.

**Did you add tests for your changes?**
Already did.

**Does this PR introduce a breaking change?**
No.

**What needs to be documented once your changes are merged?**
Nothing.

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
We need a package to parse it, because there is a lot of edge cases

**Comment by liulangyu90316** [comment]
@alexander-akait  Sorry but I do not get you, maybe I should provide a demo package or add some test cases?

**Comment by alexander-akait** [comment]
https://github.com/npm/hosted-git-info

**Comment by liulangyu90316** [comment]
@alexander-akait Thx, I know, I will add handler for more edge cases, and it will be more simple than `hosted-git-info`.

**Comment by liulangyu90316** [comment]
@alexander-akait Already add more edge cases handlers according to `hosted-git-info`. And tests can cover all the cases provided by `hosted-git-info`.

File: lib/sharing/utils.js [hunk 1]
```diff
@@ -9,13 +9,290 @@ const { join, dirname, readJson } = require("../util/fs");
...
+
+// Short url with specific protocol. eg: github:foo/bar
+const RE_GIT_URL_SHORT = /^(github|gitlab|bitbucket|gist):\/?[^/.]+\/?/;
+
+// Currently supported protocols
+const RE_PROTOCOL =
+	/^((git\+)?(ssh|https?|file)|git|github|gitlab|bitbucket|gist):$/;
+
+// Has custom protocol
+const RE_CUSTOM_PROTOCOL = /^((git\+)?(ssh|https?|file)|git):\/\//;
```
Comment (alexander-akait) [inline lib/sharing/utils.js:23]: Protocols can be `FILE://` i.e. uppercased, same for domains

**Review by alexander-akait — APPROVED** [review]

**Comment by liulangyu90316** [comment]
@alexander-akait Changes intergration and project intergration checks couldn't pass, maybe that's because I fetched some commits from main branch? Could you help giving some clue?

**Comment by buksy90** [comment]
Hi, great work, we will need support for this at my company too, but we have self hosted version of bitbucket at bitbucket.companyName.com, from a quick look at code it looks like it wont be supported.

Therefore, could you support adding custom parsers into `extractCommithashByDomain` via webpack options?
